mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java

```diff
@@ -0,0 +1,893 @@
+		if (acceptHeader == null
+				|| (!acceptHeader.contains(APPLICATION_JSON) || !acceptHeader.contains(TEXT_EVENT_STREAM))) {
+			logger.debug("Accept header validation failed. Header: {}", acceptHeader);
+			sendErrorResponse(response, "Accept header must include both application/json and text/event-stream");
+			return;
+		}
+
+		AsyncContext asyncContext = request.startAsync();
+		asyncContext.setTimeout(0);
+
+		StringBuilder body = new StringBuilder();
+		ServletInputStream inputStream = request.getInputStream();
+
+		inputStream.setReadListener(new ReadListener() {
+			@Override
+			public void onDataAvailable() throws IOException {
+				int len;
+				byte[] buffer = new byte[1024];
+				while (inputStream.isReady() && (len = inputStream.read(buffer)) != -1) {
+					body.append(new String(buffer, 0, len, StandardCharsets.UTF_8));
+				}
+			}
+
+			@Override
+			public void onAllDataRead() throws IOException {
+				try {
+					logger.debug("Parsing JSON-RPC message: {}", body.toString());
+					JSONRPCMessage message = McpSchema.deserializeJsonRpcMessage(objectMapper, body.toString());
+
+					boolean isInitializeRequest = false;
+					String sessionId = request.getHeader(SESSION_ID_HEADER);
+
+					if (message instanceof McpSchema.JSONRPCRequest req
+							&& McpSchema.METHOD_INITIALIZE.equals(req.method())) {
+						isInitializeRequest = true;
+						logger.debug("Detected initialize request");
+						if (sessionId == null) {
+							sessionId = sessionIdProvider.get();
+							logger.debug("Created new session ID for initialize request: {}", sessionId);
+						}
+					}
+
+					if (!isInitializeRequest && sessionId == null) {
+						sendErrorResponse(response, "Session ID missing in request header");
+						asyncContext.complete();
+						return;
+					}
+
+					McpServerSession session = getOrCreateSession(sessionId, isInitializeRequest);
+					if (session == null) {
+						logger.error("Failed to create session for sessionId: {}", sessionId);
+						handleSessionNotFound(sessionId, request, response);
+						asyncContext.complete();
+						return;
+					}
+
+					// Delayed until version negotiation is implemented.
+					/*
+					 * if (session.getState().equals(session.STATE_INITIALIZED) &&
+					 * request.getHeader(PROTOCOL_VERSION_HEADER) == null) {
+					 * sendErrorResponse(response,
+					 * "Protocol version missing in request header"); }
+					 */
+
+					logger.debug("Using session: {}", sessionId);
+
+					response.setHeader(SESSION_ID_HEADER, sessionId);
+
+					final String transportId;
+					if (message instanceof JSONRPCRequest req) {
+						transportId = req.id().toString();
+					}
+					else if (message instanceof JSONRPCResponse resp) {
+						transportId = resp.id().toString();
+					}
+					else {
+						transportId = null;
+					}
+
+					// Only set content type for requests
+					if (message instanceof McpSchema.JSONRPCRequest) {
+						response.setContentType(APPLICATION_JSON);
```

Comment:
So here you set it to application/json but later you let the user change this to text/event-stream. Unfortunately with other transports it might not be as straightforward and the builders are more type-safe and will forbid rolling back in such a way. That's why I am proposing to keep it simple for now as in another comment and always upgrade to SSE for requests.